presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/RuleIndex.java

```diff
@@ -12,28 +12,29 @@
  * limitations under the License.
  */
 
-package com.facebook.presto.matching;
+package com.facebook.presto.sql.planner.iterative;
 
+import com.facebook.presto.matching.Pattern;
 import com.google.common.collect.ImmutableListMultimap;
 import com.google.common.collect.ListMultimap;
 import com.google.common.reflect.TypeToken;
 
 import java.util.Set;
 import java.util.stream.Stream;
 
-public class MatchingEngine<T extends Matchable>
+public class RuleIndex
 {
-    private final ListMultimap<Class, T> matchablesByClass;
+    private final ListMultimap<Class, Rule> rulesByRootType;
```

Comment:
Let's stay with RuleIndex for now, as it:
1. fixes the [misnomer](https://github.com/prestodb/presto/pull/8376#discussion_r125303315) of MatchingEngine
2. removes the redundant `Matchables` class
3. has been modified in this PR and retrofitting the changes into `MatchingEngine` would be busywork.